### PR TITLE
implement Array1::To

### PR DESCRIPTION
**Pull request by csukuangfj**

### CODE REVIEW

**Comment by csukuangfj** [comment]
This [colab notebook][1]  shows that all tests are passed.

[1]: https://colab.research.google.com/drive/1qbHUhNZUX7AYEpqnZyf29Lrz2IPHBGlX?usp=sharing#scrollTo=U0XpYZd4_6Od

**Comment by danpovey** [comment]
Amazing!  Thanks for such fast work!
Will merge.

File: k2/csrc/dtype.h [hunk 1]
```diff
@@ -117,23 +121,58 @@ struct DtypeOf<uint64_t> {
...
-  case kUint32Dtype: { using TypeName = uint32_t; Expr; break; }\
-  case kUint64Dtype: { using TypeName = uint64_t; Expr; break; }  \
-  default: K2_FATAL << "Dtype " << TratsOf(Dtype)                \
-                    << " not covered in switch statement.  p not supported for this type?"; \
-  } while(0)
-
-
-
-
+#define FOR_ALL_DTYPES(DtypeValue, TypeName, Expr)                       \
```
Comment (danpovey) [inline k2/csrc/dtype.h:124]: mm I'm not so convinced that this is an improvement.  Is this following a particular style recommendation?

**Comment by danpovey** [comment]
anyway merged, what I queried was a small issue.

File: k2/csrc/dtype.h (hunk 1 above)
Comment (csukuangfj) [reply k2/csrc/dtype.h:124]: This is automatically formatted by `clang-format`.

**Pull request merged**